packages/flutter_tools/lib/src/commands/test.dart

```diff
@@ -419,16 +419,28 @@ class TestCommand extends FlutterCommand with DeviceBasedDevelopmentArtifacts {
       webUseWasm: useWasm,
     );
 
-    String? testAssetDirectory;
+    final Uri? nativeAssetsJson = await nativeAssetsBuilder?.build(buildInfo);
+    String? testAssetPath;
     if (buildTestAssets) {
       await _buildTestAsset(
         flavor: buildInfo.flavor,
         impellerStatus: debuggingOptions.enableImpeller,
         buildMode: debuggingOptions.buildInfo.mode,
         packageConfigPath: buildInfo.packageConfigPath,
       );
-      testAssetDirectory = globals.fs.path.
-        join(flutterProject.directory.path, 'build', 'unit_test_assets');
+    }
+    if (buildTestAssets || nativeAssetsJson != null) {
+      testAssetPath = globals.fs.path
+          .join(flutterProject.directory.path, 'build', 'unit_test_assets');
+    }
+    if (nativeAssetsJson != null) {
+      final Directory testAssetDirectory = globals.fs.directory(testAssetPath);
+      if (!await testAssetDirectory.exists()) {
```

Comment:
Should we be using the sync FS APIs here instead?

Reply:
For fast operations like `exists`, yes.